sound/soc/sof/sof-client.h

```diff
@@ -0,0 +1,69 @@
+/* SPDX-License-Identifier: (GPL-2.0 OR BSD-3-Clause) */
+/*
+ * This file is provided under a dual BSD/GPLv2 license.  When using or
+ * redistributing this file, you may do so under either license.
+ *
+ * Copyright(c) 2020 Intel Corporation. All rights reserved.
+ *
+ * Author: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>
+ */
+#ifndef __SOUND_SOC_SOF_CLIENT_H
+#define __SOUND_SOC_SOF_CLIENT_H
+
+#include <sound/sof/stream.h> /* needs to be included before control.h */
+#include <sound/sof/control.h>
+#include <sound/sof/dai.h>
+#include <sound/sof/topology.h>
+#include <sound/sof/header.h>
+#include <sound/soc.h>
+#include <sound/sof.h>
+
+enum sof_client_type {
+	SOF_CLIENT_AUDIO,
+	SOF_CLIENT_IPC,
+};
+
+struct sof_client_ops {
+	int (*probe)(struct device *dev); /* mandatory */
+	int (*remove)(struct device *dev); /* mandatory */
+
+	/* Optional IPC RX callback */
+	void (*sof_client_ipc_rx)(struct device *dev, u32 msg_cmd);
+
+	/*
+	 * Optional callback to check if the client's current status allows the
+	 * DSP to enter a low-power D0 substate when the system is in S0.
+	 */
+	bool (*allow_lp_d0_substate_in_s0)(struct device *dev);
```

Comment:
Wouldn't it make sense to make this more generic like `power_state_allowed(dev, sys_state, dev_state)`?